### PR TITLE
rmilter: Fix a couple of bugs

### PR DESCRIPTION
**Pull request by pjones**
###### Motivation for this change

  * The module uses `stringSplit` but it should be `splitString`

  * `rmilter` doesn't actually support binding to multiple sockets.
    Therefore, bind to the last one specified if `socketActivation` is
    `false`.

I also believe there is a bug in this module related to systemd
`ListenStream`.  If `socketActivation` is true, Postfix gets
connection timeouts trying to connect to one of the `ListenStream`
inet addresses.  I don't know enough about `ListenStream` passing
connections on to `fd:3` to understand what's going on.

These changes are in production (with `socketActivation = false`) via NixOps.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by avnik** [comment]
LGTM!
I will address question about socked activation to upstream author.

**Comment by Mic92** [comment]
When multiple sockets are specified systemd socket activation will pass them as file descriptors from fd:3 upwards (fd:3, fd:4, ..., fd:n)

**Comment by pjones** [comment]
> When multiple sockets are specified systemd socket activation will pass them as file descriptors from fd:3 upwards (fd:3, fd:4, ..., fd:n)

Then that also causes a problem since `rmilter` can only bind/listen to one address.

**Comment by Mic92** [comment]
If it only support one socket address at the time, the options should not allow to specify multiple. Picking the last socket does not make sense to me.

**Comment by avnik** [comment]
actually socket activation needs only because libmilter can't setup proper permissions on unix socket, but systemd can.

**Comment by avnik** [comment]
rmilter unable to listen more than one socket/port, due libmilter desing flaw

**Comment by Mic92** [comment]
socket activation is also useful to prevent race conditions between connecting smtp daemons and rmilter on startup, because systemd sockets are created before any service is started.

**Review by fpletz — CHANGES_REQUESTED** [review]
:+1: on the changes, but this will break current setups using unix sockets and `socketActivation` disabled. The sockets will have 777 permissions.

Could you please add one or two sentences to the release notes for 17.03? Thanks!

**Comment by avnik** [comment]
@fpletz It why I implement socket activation in module (and also why upstream author add it), libmilter unable to set proper permissions on socket (as well, as listen to more than one socket)

**Comment by pjones** [comment]
@avnik If rmilter can't listen on more than one socket the Nix module shouldn't allow someone to configure more than one socket.

**Comment by avnik** [comment]
On Sat, Jan 21, 2017 at 01:26:31PM -0800, Peter J. Jones wrote:

> @avnik If rmilter can't listen on more than one socket the Nix module shouldn't allow someone to configure more than one socket. 
> 

Correct.
Will try to improve it from my side (unless you want to do it as part of
this PR)

**Comment by pjones** [comment]
@avnik If you have time please go ahead and work on it.  I've got a lot on my plate right now so this is going to have to sit on the back burner for now.

**Comment by pjones** [comment]
@fpletz Maybe we should hold off on this PR until @avnik can change the interface for configuring the listening socket.  Changing the interface is going to be better than what I'm doing here.

Also, @avnik, if you come up with something let me know and I can test it.

**Review by fpletz — APPROVED** [review]

**Comment by fpletz** [comment]
I've implemented support for only one socket and deprecated the two old options. Also rebased to current master.

@avnik @pjones What do you think?

**Comment by avnik** [comment]
@fpletz  thank you. Your changes looks good.

File: nixos/modules/services/mail/rmilter.nix
Comment (pjones) [inline nixos/modules/services/mail/rmilter.nix]: Shouldn't `rspamdSocket` be `rmilterSocket` instead?

**Comment by pjones** [comment]
I added one comment, and I also think the description for the `socketActivation` option needs to be updated since it doesn't actually force rmilter to use an `inet` socket now.

File: nixos/modules/services/mail/rmilter.nix
Comment (fpletz) [inline nixos/modules/services/mail/rmilter.nix]: Oh, right! I've actually never tested this with an rspamd but will before merging of course. Thanks for reading my changes carefully! :+1:

**Comment by fpletz** [comment]
Rebased, fixed the issues @pjones mentioned and also tightened the permissions so only users in the rmilter or rspamd groups can access the unix sockets.

File: nixos/modules/services/mail/rmilter.nix
Comment (fpletz) [inline nixos/modules/services/mail/rmilter.nix]: One more thing: I removed `milter_default_action` here because it would get appended last in my `main.cf` so I couldn't overwrite it anymore. This should be set in `services.postfix.extraConfig` manually if required IMHO. The postfix default is `tempfail` which should be fine.

Comment (pjones) [inline nixos/modules/services/mail/rmilter.nix]: For what it's worth, I don't use this feature of the package.  I add the milter stuff to postfix by hand because my needs are a bit different.

**Pull request merged**
